src/Datadog.Trace/ExtensionMethods/StringExtensions.cs

```diff
@@ -69,5 +70,65 @@ public static string TrimEnd(this string value, string suffix, StringComparison
 
             return null;
         }
+
+        /// <summary>
+        /// Datadog tag requirements:
+        /// 1. Tag must start with a letter
+        /// 2. Tag cannot exceed 200 characters
+        /// 3. If the first two requirements are met, then valid characters will be retained while all other characters will be converted to underscores. Valid characters include:
+        ///    - Alphanumerics
+        ///    - Underscores
+        ///    - Minuses
+        ///    - Colons
+        ///    - Periods
+        ///    - Slashes
+        ///
+        /// Note: This method will trim leading/trailing whitespace before checking the requirements.
+        /// </summary>
+        /// <param name="value">Input string to convert into tag name</param>
+        /// <param name="normalizedTagName">If the method returns true, the normalized tag name</param>
+        /// <param name="convertPeriodsToUnderscores">A flag determining whether periods should also be converted to underscores</param>
+        /// <returns>Returns whether the conversion was successful</returns>
+        public static bool TryConvertToNormalizedTagName(this string value, out string normalizedTagName, bool convertPeriodsToUnderscores)
+        {
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                normalizedTagName = null;
+                return false;
+            }
+
+            string trimmedValue = value.Trim();
+            if (!char.IsLetter(trimmedValue[0]) || trimmedValue.Length > 200)
+            {
+                normalizedTagName = null;
+                return false;
+            }
+
+            if (convertPeriodsToUnderscores)
+            {
+                normalizedTagName = Regex.Replace(input: trimmedValue, pattern: "[^a-zA-Z0-9_:/-]", replacement: "_").ToLowerInvariant();
```

Comment:
I saw this and my first though was that if we called `ToLowerInvariant()` first, then the regex would have less work to do. (We're going to `ToLowerInvariant()` either way and I'm assuming its performance depends on the length of the string and not with how many upper-case or lower-case letters here are.)

My second that was that regex may be overkill for this relatively simple pattern, so I ran some quick benchmarks:

```
|               Method |           Job |       Runtime |        Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |-------------- |-------------- |------------:|----------:|----------:|-------:|------:|------:|----------:|
|         RegexToLower |      .NET 4.8 |      .NET 4.8 | 1,180.66 ns | 23.537 ns | 29.766 ns | 0.1869 |     - |     - |    1188 B |
|         ToLowerRegex |      .NET 4.8 |      .NET 4.8 | 1,183.25 ns | 23.517 ns | 27.082 ns | 0.1869 |     - |     - |    1188 B |
| ToLowerStringBuilder |      .NET 4.8 |      .NET 4.8 |   141.35 ns |  2.546 ns |  3.031 ns | 0.0343 |     - |     - |     217 B |
|--------------------- |-------------- |-------------- |------------:|----------:|----------:|-------:|------:|------:|----------:|
|         RegexToLower | .NET Core 3.1 | .NET Core 3.1 |   828.58 ns |  9.275 ns |  7.745 ns | 0.1335 |     - |     - |     840 B |
|         ToLowerRegex | .NET Core 3.1 | .NET Core 3.1 |   807.36 ns |  9.227 ns |  8.180 ns | 0.1335 |     - |     - |     840 B |
| ToLowerStringBuilder | .NET Core 3.1 | .NET Core 3.1 |    99.86 ns |  1.510 ns |  1.339 ns | 0.0343 |     - |     - |     216 B |
```

Looks like changing the order to `ToLowerInvariant()` only makes a tiny difference in .NET Core 3.1 and is insignificant in .NET Framework. Using a `for` look with a `StringBuilder` was definitely much master and memory-efficient, though.

```csharp
var sb = new StringBuilder(trimmedValue.ToLowerInvariant());

for (int x = 0; x < sb.Length; x++)
{
    switch (sb[x])
    {
        case >= 'a' and <= 'z' or >= '0' and <= '9' or '_' or ':' or '/' or '-':
            continue;
        default:
            sb[x] = '_';
            break;
    }
}

normalizedTagName = sb.ToString();
```

Reply:
Thanks for suggesting this new approach 😄 Running some benchmarks validates the results you got as well, so I'll make the update

Reply:
Done in b02a4fe24973e9e6be8aa48b689b30e9818d3d24